tests/test_operator.py

```diff
@@ -1186,35 +1186,78 @@ def test_equations_mixed_timedim_stepdim(self):
+        Weights are:
 
-    def test_domain_vs_interior(self):
-        """
-        Tests regions work properly in terms of code generation and runtime
-        argument derivation.
+               x=0     x=1     x=2     x=3
+        t=n     2    ---3
+                v   /
+        t=n+1   o--+----4
+
+        Flow dependency should traverse x in the negative direction
+
+               x=2     x=3     x=4     x=5      x=6
+        t=0             0   --- 0     -- 1    -- 0
+                        v  /    v    /   v   /
+        t=1            44 -+--- 11 -+--- 2--+ -- 0
         """
-        grid = Grid(shape=(4, 4, 4))
-        x, y, z = grid.dimensions
-        t = grid.stepping_dim  # noqa
+        grid = Grid(shape=(10, 10))
+        x, y = grid.dimensions
+        u = TimeFunction(name='u', grid=grid, save=2, time_order=1, space_order=0)
+        step = Eq(u.forward, 2*u
+                  + 3*u.subs(x, x+x.spacing)
+                  + 4*u.forward.subs(x, x+x.spacing))
+        op = Operator(step)
 
-        u = TimeFunction(name='u', grid=grid)  # noqa
-        eqs = [Eq(u.forward, u + 1),
-               Eq(u.forward, u.forward + 2, region=INTERIOR)]
+        u.data[:] = 0.0
+        u.data[0, 5, 5] = 1.0
+
+        op.apply(time_M=0)
+        assert u.data[1, 5, 5] == 2
+        assert u.data[1, 4, 5] == 11
+        assert u.data[1, 3, 5] == 44
+        assert u.data[1, 2, 5] == 4*44
+        assert u.data[1, 1, 5] == 4*4*44
+        assert u.data[1, 0, 5] == 4*4*4*44
+        assert np.all(u.data[1, 6:, :] == 0)
+        assert np.all(u.data[1, :, 0:5] == 0)
+        assert np.all(u.data[1, :, 6:] == 0)
+
+    def test_scheduling_sparse_functions(self):
+        """Tests loop scheduling in presence of sparse functions."""
+        grid = Grid((10, 10))
+        time = grid.time_dim
 
-        op = Operator(eqs, dse='noop', dle='noop')
-        trees = retrieve_iteration_tree(op)
-        assert len(trees) == 2
+        u1 = TimeFunction(name="u1", grid=grid, save=10, time_order=2)
+        u2 = TimeFunction(name="u2", grid=grid, time_order=2)
+        sf1 = SparseFunction(name='sf1', grid=grid, npoint=1, ntime=10)
+        sf2 = SparseFunction(name='sf2', grid=grid, npoint=1, ntime=10)
 
-        op.apply(time_M=1)
-        assert np.all(u.data[1, 0, :, :] == 1)
-        assert np.all(u.data[1, -1, :, :] == 1)
-        assert np.all(u.data[1, :, 0, :] == 1)
-        assert np.all(u.data[1, :, -1, :] == 1)
-        assert np.all(u.data[1, :, :, 0] == 1)
-        assert np.all(u.data[1, :, :, -1] == 1)
-        assert np.all(u.data[1, 1:3, 1:3, 1:3] == 3)
+        # Deliberately inject into u1, rather than u1.forward, to create a WAR
+        eqn1 = Eq(u1.forward, u1 + 2.0 - u1.backward)
+        eqn2 = sf1.inject(u1, expr=sf1)
+        eqn3 = Eq(u2.forward, u2 + 2*u2.backward - u1.dt2)
+        eqn4 = sf2.interpolate(u2)
+
+        op = Operator([eqn1] + eqn2 + [eqn3] + eqn4)
+        trees = retrieve_iteration_tree(op)
```

Comment:
```
        u1 = TimeFunction(name="u1", grid=grid, save=10, time_order=2)
        u2 = TimeFunction(name="u2", grid=grid, time_order=2)
        const = Function(name="const", grid=grid, space_order=2)
        sf1 = SparseFunction(name='sf1', grid=grid, npoint=1, ntime=10)
        sf2 = SparseFunction(name='sf2', grid=grid, npoint=1, ntime=10)

        # Deliberately inject into u1, rather than u1.forward, to create a WAR
        from sympy import cos
        eqn1 = Eq(u1.forward, u1 + 2.0 - u1.backward + cos(const))
        eqn2 = sf1.inject(u1.forward, expr=sf1)
        eqn3 = Eq(u2.forward, u2 + 2*u2.backward - u1.dt2 + cos(const))
        eqn4 = sf2.interpolate(u2)

        op = Operator([eqn1] + eqn2 + [eqn3] + eqn4)
```

This one is slightly more complicated, and can be part of some other PR but creates two issues:
- time invariants are redeclared with same name for different clusters
- Time loop is interupted by the second time invariant and restarted

Reply:
I suspect we can produce a smaller failing example without using ``SparseFunction``. Will spend some time on this later this afternoon